File: python/섬의 개수(4963) copy.py
from collections import deque 
import sys
input = sys.stdin.readline

while True:
    w, h = tuple(map(int, input().split()))

    if w == 0 and h == 0:
        break

    graph = [
        list(map(int, input().split()))
        for _ in range(h)
    ]

    visited = [[False] * w for _ in range(h)]

    dx = [-1,1,0,0,1,1,-1,-1]
    dy = [0,0,-1,1,-1,1,-1,1]

    def bfs(x,y):
        queue = deque()
        queue.append((x,y))
        visited[x][y] = True

        while queue:
            x,y = queue.popleft()
            for i in range(8):
                nx = x + dx[i]
                ny = y + dy[i]

                if 0 <= nx < h and 0 <= ny < w and not visited[nx][ny] and graph[nx][ny] == 1:
                    queue.append((nx,ny))
                    visited[nx][ny] = True
        
    cnt = 0
    for i in range(h):
        for j in range(w):
            if not visited[i][j] and graph[i][j] == 1:
                bfs(i,j)
                cnt += 1
    
    print(cnt)

File: js/섬의 개수(4963) copy.js
const input = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n');

let index = 0;
while (true) {
    const [w,h] = input[index].split(' ').map(Number);
    index++;

    if (w === 0 && h === 0) break;

    const graph = input.slice(index, index+h).map((line) => line.split(' ').map(Number));
    const visited = Array.from({length:h}, () => Array(w).fill(false));
    index += h;

    const dx = [-1, 1, 0, 0, -1, -1, 1, 1];
    const dy = [0, 0, -1, 1, -1, 1, -1, 1];

    function bfs(x,y) {
        let queue = [];
        queue.push([x,y]);
        visited[x][y] = true;

        while (queue.length) {
            const [x,y] = queue.shift();

            for (let i=0; i<8; i++) {
                let nx = x + dx[i];
                let ny = y + dy[i];

                if (0 <= nx && nx < h && 0 <= ny && ny < w && !visited[nx][ny] && graph[nx][ny] === 1) {
                    queue.push([nx,ny]);
                    visited[nx][ny] = true;
                }
            }
        }
    }

    let cnt = 0;
    for (let i=0; i<h; i++) {
        for (let j=0; j < w; j++) {
            if (graph[i][j] === 1 && !visited[i][j]) {
                bfs(i,j);
                cnt++;
            }
        }
    }

    console.log(cnt);
}